README.md
# Golang ISO-8583

ISO-8583 support library for Golang

## Usage

```
// Encode
m := iso8583.Message{
  DE2:   NewNumeric("4846811212"),        // Primary Account Number
  DE3:   NewNumeric("201234"),            // Processing Code
  DE4:   NewNumeric("10000000"),          // Amount, Transaction
  DE7:   NewNumeric("1107221800"),        // Date And Time, Transmission
  DE11:  NewNumeric("000001"),            // Systems Trace Audit Number
  DE12:  NewNumeric("161204171926"),      // Date And Time, Local Transaction
  DE22:  NewAlphanumeric("FABCDE123ABD"), // Point Of Service Data Code
  DE32:  NewNumeric("414243"),            // Acquiring Institution Identification Code
  DE39:  NewNumeric("000"),               // Action Code
  DE41:  NewANS("termid12"),              // Card Acceptor Terminal Identification
  DE43:  NewANS("Community1"),            // Card Acceptor Name/Location
  DE102: NewANS("12341234234"),           // Account Identification 1
}
m.Mti = "1200"
m.SetEncoder(iso8583.ASCII)

b, _ := m.Encode()
//  Output: 1200F230040102A0000000000000040000001048468112122012340000100000001107221800000001161204171926FABCDE123ABD06414243000termid1210Community11112341234234
```

```
// Decode
msg := "1210FA304555AAE4840600000000100000001600000000000000000920000000000000000000000000000123205001030402950123154952591221010121314C2005912950123000000020000000000020000111007640125111101111111182954212248887288158=99120010109012401      002NJ02011173420          58NJ NEWARK          123 PINE STREET      USWRIGHT AID DRUGS0030008400230202017840D00000001500077700101231110222222226"
	m := Message{}
	m.SetEncoder(iso8583.ASCII)
	_ := m.Decode([]byte(msg))
```


CHANGELOG.md
## Changelog

**0.3.0 - 2020 Jun 18**

- message
	- add DE125 as `*SubMessage` type

- fields
	- rename `B` field type to `B64`, because it represents a 64 length `B`
	- add `BN` as a new field, because it represents an N length `B`
	- add `Reserved`, it throws error in case of usage of reserved fields

```go
/ Usage of submessage in 0.3.0
m := &Message{
	DE125: &SubMessage{
		SE2: NewANS("Test Address"),
	},
}
result, _ := m.Encode() // err handle
```

**0.2.0 - 2020 Jun 17**

- submessage
	- add submessage which is the field representation of DE125
	- encode
	- decode

```go
// Usage of submessage in 0.2.0
sm := &SubMessage{
	SE2: NewANS("Test Address"),
}
b, _ := sm.Encode() // err handle

m := &Message{
	DE125: NewANS(string(b)),
}
result, _ := m.Encode() // err handle
```

**0.1.0 - 2020 Jun 16**

- message
	- definition of iso8583 base structure
	- encode
	- decode
	- bitmap related functions

- fields
	- definition of field types of iso8583
	- regexp to validate fields

mti_test.go
package iso8583

import (
	"bytes"
	"testing"
)

func TestMti(t *testing.T)  {
	b,err := encodeMti("1100")
	if err != nil {
		t.Fatal(err)
	}
	bytes.Equal(b,[]byte("1100"))
}


bitmap_test.go
package iso8583

import (
	"reflect"
	"testing"
)

func TestAddField(t *testing.T) {
	var scenarios = []struct {
		result string
		bits   []uint8
	}{
		{
			result: "42000104021C0078",
			bits:   []uint8{2, 7, 24, 30, 39, 44, 45, 46, 58, 59, 60, 61},
		},
		{
			result: "4200010002140068",
			bits:   []uint8{2, 7, 24, 39, 44, 46, 58, 59, 61},
		},
		{
			result: "0000010002140068",
			bits:   []uint8{24, 39, 44, 46, 58, 59, 61},
		},
		{
			result: "A888010002143068",
			bits:   []uint8{1, 3, 5, 9, 13, 24, 39, 44, 46, 51, 52, 58, 59, 61},
		},
	}

	for _, scenario := range scenarios {
		var fields uint64
		for _, bit := range scenario.bits {
			fields = addField(fields, bit)
		}
		if result := bitmapHex(fields); scenario.result != result {
			t.Errorf("bitmap should be %s, instead of %s", scenario.result, result)
		}
	}
}

func TestIsBitSet(t *testing.T) {
	var scenarios = []struct {
		bits []uint8
	}{
		{
			bits: []uint8{2, 7, 24, 30, 39, 44, 45, 46, 58, 59, 60, 61},
		},
		{
			bits: []uint8{2, 7, 24, 39, 44, 46, 58, 59, 61},
		},
		{
			bits: []uint8{24, 39, 44, 46, 58, 59, 61},
		},
		{
			bits: []uint8{1, 3, 5, 9, 13, 24, 39, 44, 46, 51, 52, 58, 59, 61},
		},
		{
			bits: []uint8{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 61, 62, 63, 64},
		},
	}

	for _, scenario := range scenarios {
		var fields uint64
		// create bitset
		for _, bit := range scenario.bits {
			fields = addField(fields, bit)
		}

		bits := make([]uint8, 0)
		for i := uint8(1); i <= 64; i++ {
			if isBitSet(fields, i) {
				bits = append(bits, i)
			}
		}

		if !reflect.DeepEqual(scenario.bits, bits) {
			t.Errorf("bitmap should be %#v, instead of %#v", scenario.bits, bits)
		}
	}
}

func TestDecodeHexString(t *testing.T) {
	var scenarios = []struct {
		bitmapString string
		bits         []uint8
	}{
		{
			bitmapString: "42000104021C0078",
			bits:         []uint8{2, 7, 24, 30, 39, 44, 45, 46, 58, 59, 60, 61},
		},
		{
			bitmapString: "4200010002140068",
			bits:         []uint8{2, 7, 24, 39, 44, 46, 58, 59, 61},
		},
		{
			bitmapString: "0000010002140068",
			bits:         []uint8{24, 39, 44, 46, 58, 59, 61},
		},
		{
			bitmapString: "A888010002143068",
			bits:         []uint8{1, 3, 5, 9, 13, 24, 39, 44, 46, 51, 52, 58, 59, 61},
		},
		{
			bitmapString: "FFFFFF000000000F",
			bits:         []uint8{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 61, 62, 63, 64},
		},
	}

	for _, scenario := range scenarios {
		var expected uint64
		for _, bit := range scenario.bits {
			expected = addField(expected, bit)
		}

		if result, err := decodeHexString(scenario.bitmapString); err != nil {
			t.Error(err)
		} else if expected != result {
			t.Errorf("bitmap should be %d, instead of %d", expected, result)
		}
	}
}


fields_test.go
package iso8583

import (
	"bytes"
	"testing"
)

func TestNumericFixed(t *testing.T) {
	n := NewNumeric("12345")
	b, err := n.Encode(ASCII, 5, "", "N")
	if err != nil {
		t.Fatal(err)
	}
	if !bytes.Equal(b, []byte("12345")) {
		t.Error("bad encoding")
	}

	n = NewNumeric("12345")
	b, err = n.Encode(ASCII, 6, "", "N")
	if err != nil {
		t.Fatal(err)
	}
	if !bytes.Equal(b, []byte("012345")) {
		t.Error("bad encoding")
	}

	n = NewNumeric("12345")
	b, err = n.Encode(ASCII, 10, "", "N")
	if err != nil {
		t.Fatal(err)
	}
	if !bytes.Equal(b, []byte("0000012345")) {
		t.Error("bad encoding")
	}
}

func TestNumericLLVAR(t *testing.T) {
	n := NewNumeric("12345")
	b, err := n.Encode(ASCII, 19, "LLVAR", "N")
	if err != nil {
		t.Fatal(err)
	}
	if !bytes.Equal(b, []byte("0512345")) {
		t.Error("bad encoding")
	}

	b, err = n.Encode(ASCII, 4, "LLVAR", "N")
	if err == nil {
		t.Error("expecting error, length 4 < len(n)")
	}

	b, err = n.Encode(ASCII, 5, "LLVAR", "N")
	if err != nil {
		t.Error(err)
	}
}

func TestNumericLLLVAR(t *testing.T) {
	n := NewNumeric("12345")
	b, err := n.Encode(ASCII, 19, "LLLVAR", "N")
	if err != nil {
		t.Fatal(err)
	}
	if !bytes.Equal(b, []byte("00512345")) {
		t.Error("bad encoding")
	}

	b, err = n.Encode(ASCII, 4, "LLLVAR", "N")
	if err == nil {
		t.Error("expecting error, length 4 < len(n)")
	}

	b, err = n.Encode(ASCII, 5, "LLLVAR", "N")
	if err != nil {
		t.Error(err)
	}
}

func TestAlphaNumericFixed(t *testing.T) {
	n := NewAlphanumeric("12AN")
	b, err := n.Encode(ASCII, 10, "", "AN")
	if err != nil {
		t.Fatal(err)
	}
	if !bytes.Equal(b, []byte("12AN      ")) {
		t.Error("bad encoding")
	}
}

func TestAlphaNumericLLVAR(t *testing.T) {
	n := NewAlphanumeric("12ANABCDEFGHTCASDASSAASCSACSACSACSACSACACSACSACS")
	b, err := n.Encode(ASCII, 99, "LLVAR", "AN")
	if err != nil {
		t.Fatal(err)
	}
	if !bytes.Equal(b, []byte("4812ANABCDEFGHTCASDASSAASCSACSACSACSACSACACSACSACS")) {
		t.Error("bad encoding")
	}
}

func TestAlphaNumericLLLVAR(t *testing.T) {
	n := NewAlphanumeric("12ANABCDEFGHTCASDASSAASCSACSACSACSACSACACSACSACS")
	b, err := n.Encode(ASCII, 99, "LLLVAR", "AN")
	if err != nil {
		t.Fatal(err)
	}
	if !bytes.Equal(b, []byte("04812ANABCDEFGHTCASDASSAASCSACSACSACSACSACACSACSACS")) {
		t.Error("bad encoding")
	}
}

func TestNDecode(t *testing.T) {
	b := []byte("0512345")
	n := NewNumeric("")
	_, err := n.Decode(b, ASCII, 19, "LLVAR", "N")
	if err != nil {
		t.Fatal(err)
	}
	t.Log(string(n.Value))

	b = []byte("0000012356")
	nn, err := n.Decode(b, ASCII, 10, "", "")
	if err != nil {
		t.Fatal(err)
	}
	t.Log(string(n.Value))
	t.Log(nn)
}


mti.go
package iso8583

import "errors"

const (
	AuthorizationRequest         = "1100"
	AuthorizationRequestResponse = "1110"
	AuthorizationAdvice          = "1120"
	AuthorizationAdviceRepeat    = "1121"
	AuthorizationAdviceResponse  = "1130"

	FinancialRequest         = "1200"
	FinancialRequestResponse = "1210"
	FinancialAdvice          = "1220"
	FinancialAdviceRepeat    = "1221"
	FinancialAdviceResponse  = "1230"

	FileActionRequest  = "1304"
	FileActionResponse = "1314"

	ReversalAdvice         = "1420"
	ReversalAdviceRepeat   = "1421"
	ReversalAdviceResponse = "1430"

	AdministrativeAdvice         = "1624"
	AdministrativeAdviceResponse = "1634"

	NetworkManagementRequest         = "1804"
	NetworkManagementRequestResponse = "1814"
)

var mti = map[string]struct{}{
	AuthorizationRequest:         {},
	AuthorizationRequestResponse: {},
	AuthorizationAdvice:          {},
	AuthorizationAdviceRepeat:    {},
	AuthorizationAdviceResponse:  {},

	FinancialRequest:         {},
	FinancialRequestResponse: {},
	FinancialAdvice:          {},
	FinancialAdviceRepeat:    {},
	FinancialAdviceResponse:  {},

	FileActionRequest:  {},
	FileActionResponse: {},

	ReversalAdvice:         {},
	ReversalAdviceRepeat:   {},
	ReversalAdviceResponse: {},

	AdministrativeAdvice:         {},
	AdministrativeAdviceResponse: {},

	NetworkManagementRequest:         {},
	NetworkManagementRequestResponse: {},
}

func encodeMti(messageTypeId string) ([]byte, error) {
	if _, ok := mti[messageTypeId]; !ok {
		return nil, errors.New("invalid message type identifier")
	}
	return []byte(messageTypeId), nil
}


message.go
package iso8583

import (
	"errors"
	"reflect"
	"strconv"
	"strings"
)

type Message struct {
	Mti string

	packedBitmap bool
	packedMsg    bool
	encoder      int

	bitmapPrimary uint64

	DE1   uint64      `format:"" length:"64"` //secondary bitmap
	DE2   *N          `format:"LLVAR" length:"19" validator:"N"`
	DE3   *N          `format:"" length:"6" validator:"N"`
	DE4   *N          `format:"" length:"12" validator:"N"`
	DE5   *N          `format:"" length:"12" validator:"N"`
	DE6   *N          `format:"" length:"12" validator:"N"`
	DE7   *N          `format:"" length:"10" validator:"MMDDHHMMSS"`
	DE9   *N          `format:"" length:"8" validator:"N"`
	DE10  *N          `format:"" length:"8" validator:"N"`
	DE11  *N          `format:"" length:"6" validator:"N"`
	DE12  *N          `format:"" length:"12" validator:"YYMMDDHHMMSS"`
	DE14  *N          `format:"" length:"4" validator:"YYMM"`
	DE16  *N          `format:"" length:"4" validator:"MMDD"`
	DE17  *N          `format:"" length:"4" validator:"MMDD"`
	DE18  *N          `format:"" length:"4" validator:"N"`
	DE19  *N          `format:"" length:"3" validator:"N"`
	DE22  *AN         `format:"" length:"12" validator:"AN"`
	DE23  *N          `format:"" length:"3" validator:"N"`
	DE24  *N          `format:"" length:"3" validator:"N"`
	DE25  *N          `format:"" length:"4" validator:"N"`
	DE26  *N          `format:"" length:"4" validator:"N"`
	DE28  *N          `format:"" length:"6" validator:"YYMMDD"`
	DE30  *N          `format:"" length:"24" validator:"N"`
	DE32  *N          `format:"LLVAR" length:"11" validator:"N"`
	DE33  *N          `format:"LLVAR" length:"11" validator:"N"`
	DE34  *N          `format:"LLVAR" length:"28" validator:"N"`
	DE35  *Z          `format:"LLVAR" length:"37" validator:"Z"`
	DE37  *ANP        `format:"" length:"12" validator:"ANP"`
	DE38  *ANP        `format:"" length:"6" validator:"ANP"`
	DE39  *N          `format:"" length:"3" validator:"N"`
	DE41  *ANS        `format:"" length:"8" validator:"ANS"`
	DE42  *ANS        `format:"" length:"15" validator:"ANS"`
	DE43  *ANS        `format:"LLVAR" length:"99" validator:"ANS"`
	DE46  *ANS        `format:"LLLVAR" length:"186" validator:"ANS"`
	DE47  *ANS        `format:"LLLVAR" length:"999" validator:"ANS"`
	DE48  *ANS        `format:"LLLVAR" length:"999" validator:"ANS"`
	DE49  *N          `format:"" length:"3" validator:"N"`
	DE50  *N          `format:"" length:"3" validator:"N"`
	DE51  *N          `format:"" length:"3" validator:"N"`
	DE52  *B64        `format:"" length:"64" validator:"B64"`
	DE54  *ANS        `format:"LLLVAR" length:"120" validator:"ANS"`
	DE56  *N          `format:"LLVAR" length:"35" validator:"N"`
	DE57  *N          `format:"" length:"3" validator:"N"`
	DE58  *N          `format:"LLVAR" length:"11" validator:"N"`
	DE59  *ANS        `format:"LLLVAR" length:"999" validator:"ANS"`
	DE62  *N          `format:"" length:"6" validator:"N"`
	DE63  *N          `format:"" length:"4" validator:"MMDD"`
	DE66  *ANS        `format:"LLLVAR" length:"204" validator:"ANS"`
	DE72  *ANS        `format:"LLLVAR" length:"999" validator:"ANS"`
	DE93  *N          `format:"LLVAR" length:"11" validator:"N"`
	DE94  *N          `format:"LLVAR" length:"11" validator:"N"`
	DE95  *ANS        `format:"LLLLVAR" length:"9999" validator:"ANS"`
	DE96  *ANS        `format:"LLLVAR" length:"100" validator:"ANS"`
	DE100 *N          `format:"LLVAR" length:"11" validator:"N"`
	DE101 *ANS        `format:"LLVAR" length:"17" validator:"ANS"`
	DE102 *ANS        `format:"LLVAR" length:"28" validator:"ANS"`
	DE103 *ANS        `format:"LLVAR" length:"28" validator:"ANS"`
	DE111 *ANS        `format:"LLLLVAR" length:"9999" validator:"ANS"`
	DE123 *ANS        `format:"LLLVAR" length:"999" validator:"ANS"`
	DE124 *ANS        `format:"LLLVAR" length:"999" validator:"ANS"`
	DE125 *SubMessage `format:"LLLVAR" length:"999" validator:"ANS"`
	DE126 *ANS        `format:"LLLVAR" length:"999" validator:"ANS"`
	DE127 *ANS        `format:"LLLLVAR" length:"9999" validator:"ANS"`
	DE128 *ANS        `format:"LLLLLVAR" length:"99999" validator:"ANS"`
}

func (m *Message) Encode() ([]byte, error) {
	res := make([]byte, 0)

	// append mti
	if len(m.Mti) != 4 {
		return []byte{}, errors.New("invalid MTI length")
	}
	switch m.encoder {
	case BCDIC:
	case ASCII:
		res = append(res, []byte(m.Mti)...)
	}

	// initialize bitmaps
	var bitmapPrimary uint64
	var bitmapSecondary uint64

	data := make([]byte, 0, 512)

	v := reflect.Indirect(reflect.ValueOf(m))
	t := v.Type()
	// iterate through iso fields, if field is not empty,
	// encode and append it to data, and set the proper bit in bitmap
	for i := 0; i < v.NumField(); i++ {
		if v.Field(i).Kind() != reflect.Ptr || v.Field(i).IsNil() {
			continue
		}
		sf := t.Field(i)
		// get field index, e.g. for DE2 index=2
		index, err := strconv.Atoi(strings.Trim(sf.Name, "DE"))
		if err != nil {
			return nil, err
		}
		// field length
		length, err := strconv.Atoi(sf.Tag.Get("length"))
		if err != nil {
			return nil, err
		}
		// iso field format, "" means that field has fixed length, but
		// e.g. LLVAR means that length indicator is encoded in the first 2 byte
		// in this case length means "maximum length
		format := sf.Tag.Get("format")
		validator := sf.Tag.Get("validator")

		if index <= 64 {
			bitmapPrimary = addField(bitmapPrimary, uint8(index))
		} else {
			// if we need secondary bitmap, set first bit in primary bitmap
			bitmapPrimary |= 1 << 63
			bitmapSecondary = addField(bitmapSecondary, uint8(index-64))
		}

		// tweak of submessage
		f, ok := v.Field(i).Interface().(field)
		if !ok {
			if sm, ok := v.Field(i).Interface().(*SubMessage); ok {
				res, err := sm.Encode()
				if err != nil {
					return nil, err
				}
				d, err := NewANS(string(res)).Encode(m.encoder, length, format, validator)
				data = append(data, d...)
				continue
			} else {
				continue
			}
		}

		// encode field, append it to data
		d, err := f.Encode(m.encoder, length, format, validator)
		if err != nil {
			return nil, err
		}
		data = append(data, d...)
	}

	// append bitmaps to result
	m.bitmapPrimary = bitmapPrimary
	res = append(res, []byte(bitmapHex(bitmapPrimary))...)

	if bitmapSecondary != 0 {
		m.DE1 = bitmapSecondary
		res = append(res, []byte(bitmapHex(bitmapSecondary))...)
	}

	// append iso data elements to result
	res = append(res, data...)
	return res, nil
}

func (m *Message) Decode(bytes []byte) error {
	var err error

	// decode MTI

	// it is an iterator, watching where we are currently in the iteration,
	// which byte will be the starting position of the next decode
	it := 4
	m.Mti = string(bytes[:it])

	// decode bitmaps
	//decode primary bitmap
	m.bitmapPrimary, err = decodeHexString(string(bytes[it : it+16]))
	if err != nil {
		return err
	}
	it += 16

	// if first bit is 1, it means that we have secondary bitmap, decode secondary bitmap
	if isBitSet(m.bitmapPrimary, 1) {
		m.DE1, err = decodeHexString(string(bytes[it : it+16]))
		if err != nil {
			return err
		}
		it += 16
	}

	v := reflect.Indirect(reflect.ValueOf(m))
	t := v.Type()
	// iterate through iso fields, if bitmap is not empty at bit position i,
	// set field with index i with proper value
	for i := 0; i < v.NumField(); i++ {
		if v.Field(i).Kind() != reflect.Ptr {
			continue
		}
		sf := t.Field(i)
		// get field index, e.g. for DE2 index=2
		index, err := strconv.Atoi(strings.Trim(sf.Name, "DE"))
		if err != nil {
			return err
		}
		// if index < 64, search in primary bitmap if it is set
		if index <= 64 && !isBitSet(m.bitmapPrimary, uint8(index)) {
			continue
		}

		// if index > 64, search in secondary bitmap if it is set
		if index > 64 && !isBitSet(m.DE1, uint8(index-64)) {
			continue
		}

		// field (maximum) length
		length, err := strconv.Atoi(sf.Tag.Get("length"))
		if err != nil {
			return err
		}
		// iso field format, "" means that field has fixed length, but
		// e.g. LLVAR means that length indicator is encoded in the first 2 byte
		// in this case length means "maximum length
		format := sf.Tag.Get("format")
		validator := sf.Tag.Get("validator")

		// Decode field
		structField := v.Field(i)

		// skip if unexported
		if !structField.CanSet() {
			continue
		}
		// initialize field with empty struct
		fieldTyp := reflect.New(structField.Type().Elem())
		structField.Set(fieldTyp)

		var nextFieldOffset int
		f, ok := v.Field(i).Interface().(field)
		if !ok {
			if sm, ok := v.Field(i).Interface().(*SubMessage); ok {
				ans := NewANS("")
				nextFieldOffset, err = ans.Decode(bytes[it:], m.encoder, length, format, validator)
				err = sm.Decode(ans.Value)
				if err != nil {
					return err
				}
				continue
				//v.Field(i).Set(reflect.ValueOf(sm))
			} else {
				continue
			}
		}
		nextFieldOffset, err = f.Decode(bytes[it:], m.encoder, length, format, validator)
		if err != nil {
			return err
		}
		it += nextFieldOffset
	}
	return nil
}

func (m *Message) PackedBitmap(packed bool) {
	m.packedBitmap = packed
}

func (m *Message) PackedMessage(packed bool) {
	m.packedMsg = packed
}

func (m *Message) SetEncoder(encoder int) {
	m.encoder = encoder
}


regexes.go
package iso8583

import "regexp"

const (
	binary64RegexString     = "^[0-1]{64}|[0-9A-F]{16}|[0-9A-F]{8}$"
	binaryRegexString       = "^[0-9A-F]+$"
	numberRegexString       = "^[0-9]+$"
	alphaNumericRegexString = "^[a-zA-Z0-9]+$"
	anpRegexString          = "^[a-zA-Z0-9]*\\s*$"
	ansRegexString          = "^[ -~]*$"
	yymmddhhmmssRegexString = "^([0-9]{2})(0[1-9]|1[0-2])(0[1-9]|[1-2][0-9]|3[0-1])([0-1][0-9]|2[0-3])([0-5][0-9]){2}$"
	mmddhhmmssRegexString   = "^(0[1-9]|1[0-2])(0[1-9]|[1-2][0-9]|3[0-1])([0-1][0-9]|2[0-3])([0-5][0-9]){2}$"
	yymmRegexString         = "^([0-9]{2})(0[1-9]|1[0-2])$"
	mmddRegexString         = "^^(0[1-9]|1[0-2])(0[1-9]|[1-2][0-9]|3[0-1])$"
	yymmddRegexString       = "^([0-9]{2})(0[1-9]|1[0-2])(0[1-9]|[1-2][0-9]|3[0-1])$"
	track2RegexString       = "^[0-9=D]+$"
)

var (
	binary64Regex     = regexp.MustCompile(binary64RegexString)
	binaryRegex       = regexp.MustCompile(binaryRegexString)
	numberRegex       = regexp.MustCompile(numberRegexString)
	alphaNumericRegex = regexp.MustCompile(alphaNumericRegexString)
	anpRegex          = regexp.MustCompile(anpRegexString)
	ansRegex          = regexp.MustCompile(ansRegexString)
	yymmddhhmmssRegex = regexp.MustCompile(yymmddhhmmssRegexString)
	mmddhhmmssRegex   = regexp.MustCompile(mmddhhmmssRegexString)
	yymmRegex         = regexp.MustCompile(yymmRegexString)
	mmddRegex         = regexp.MustCompile(mmddRegexString)
	yymmddRegex       = regexp.MustCompile(yymmddRegexString)
	track2Regex       = regexp.MustCompile(track2RegexString)
)


submessage.go
package iso8583

import (
	"reflect"
	"strconv"
	"strings"
)

type SubMessage struct {
	encoder       int
	bitmapPrimary uint64

	SE1 uint64 `format:"" length:"64"`
	SE2 *ANS   `format:"" length:"29" validator:"ANS"`
	SE3 *ANS   `format:"" length:"5" validator:"ANS"`
	SE4 *N     `format:"" length:"10" validator:"N"`
	SE5 *Reserved
	SE6 *Reserved
	SE7 *ANS `format:"" length:"3" validator:"ANS"`
	SE8 *ANS `format:"" length:"1" validator:"ANS"`
	SE9 *AN  `format:"" length:"3" validator:"AN"`
	SE10 *Reserved
	SE11 *AN `format:"" length:"1" validator:"AN"`
	SE12 *Reserved
	SE13 *Reserved
	SE14 *Reserved
	SE15 *ANS `format:"" length:"2" validator:"ANS"`
	SE16 *ANS `format:"" length:"2" validator:"ANS"`
	SE17 *Reserved
	SE18 *ANS `format:"" length:"16" validator:"ANS"`
	SE19 *ANS `format:"" length:"999" validator:"ANS"`
	SE20 *ANS `format:"" length:"2" validator:"ANS"`
	SE21 *ANS `format:"" length:"194" validator:"ANS"` // ANS 255 if Tagged field 0008 is included
	SE22 *ANS `format:"LLLVAR" length:"255" validator:"ANS"`
	SE23 *Reserved
	SE24 *ANS `format:"LLVAR" length:"99" validator:"ANS"`
	SE25 *ANS `format:"LLVAR" length:"99" validator:"ANS"`
	SE26 *N   `format:"" length:"3" validator:"N"`
	SE27 *ANS `format:"" length:"1" validator:"ANS"`
	SE28 *Reserved
	SE29 *ANS `format:"" length:"9" validator:"ANS"`
	SE30 *N   `format:"" length:"4" validator:"N"`
	SE31 *ANS `format:"LLLVAR" length:"255" validator:"ANS"`
	SE32 *Reserved
	SE33 *Reserved
	SE34 *N   `format:"" length:"11" validator:"N"`
	SE35 *N   `format:"" length:"11" validator:"N"`
	SE36 *ANS `format:"" length:"15" validator:"ANS"`
	SE37 *AN  `format:"" length:"7" validator:"AN"`
	SE38 *ANP `format:"" length:"15" validator:"ANP"`
	SE39 *AN  `format:"LLLVAR" length:"120" validator:"AN"`
	SE40 *N   `format:"" length:"2" validator:"N"`
	SE41 *BN  `format:"LLLVAR" length:"100" validator:"BN"`
	SE42 *N   `format:"" length:"3" validator:"N"`
	SE43 *Reserved
	SE44 *Reserved
	SE45 *Reserved
	SE46 *Reserved
	SE47 *Reserved
	SE48 *Reserved
	SE49 *Reserved
	SE50 *Reserved
	SE51 *Reserved
	SE52 *Reserved
	SE53 *Reserved
	SE54 *ANS `format:"" length:"6" validator:"ANS"`
	SE55 *AN  `format:"LLLVAR" length:"120" validator:"AN"`
	SE56 *N   `format:"LLVAR" length:"45" validator:"N"`
	SE57 *AN  `format:"LLVAR" length:"10" validator:"AN"`
	SE58 *N   `format:"LLVAR" length:"30" validator:"N"`
	SE59 *AN  `format:"LLVAR" length:"10" validator:"AN"`
	SE60 *N   `format:"" length:"1" validator:"N"`
	SE61 *N   `format:"" length:"4" validator:"N"`
	SE62 *N   `format:"" length:"1" validator:"N"`
	SE63 *ANS `format:"" length:"94" validator:"ANS"`
	SE64 *AN  `format:"" length:"1" validator:"AN"`
	SE65 *AN  `format:"" length:"1" validator:"AN"`
	SE66 *AN  `format:"" length:"6" validator:"AN"`
	SE67 *AN  `format:"" length:"1" validator:"AN"`
	SE68 *ANS `format:"" length:"40" validator:"ANS"`
	SE69 *ANP `format:"" length:"6" validator:"ANP"`
	SE70 *ANS `format:"" length:"15" validator:"ANS"`
	SE71 *ANS `format:"LLLVAR" length:"999" validator:"ANS"`
	SE72 *ANS `format:"LLLVAR" length:"999" validator:"ANS"`
	SE73 *ANS `format:"LLLVAR" length:"999" validator:"ANS"`
	SE74 *ANS `format:"LLLVAR" length:"999" validator:"ANS"`
	SE75 *ANS `format:"" length:"3" validator:"ANS"`
	SE76 *ANS `format:"" length:"23" validator:"ANS"`
	SE77 *ANS `format:"" length:"12" validator:"ANS"`
	SE78 *ANS `format:"" length:"15" validator:"ANS"`
	SE79 *ANS `format:"" length:"4" validator:"ANS"`
	SE80 *ANS `format:"" length:"1" validator:"ANS"`
	SE81 *ANS `format:"" length:"2" validator:"ANS"`
	SE82 *ANS `format:"" length:"1" validator:"ANS"`
	SE83 *AN  `format:"" length:"1" validator:"AN"`
	SE84 *ANS `format:"" length:"11" validator:"ANS"`
	SE85 *BN  `format:"LLLVAR" length:"256" validator:"BN"`
	SE86 *ANS `format:"" length:"1" validator:"ANS"`
	SE87 *Reserved
	SE88 *ANS `format:"" length:"1" validator:"ANS"`
	SE89 *ANS `format:"" length:"15" validator:"ANS"`
	SE90 *ANS `format:"" length:"6" validator:"ANS"`
	SE91 *ANS `format:"LLLVAR" length:"255" validator:"ANS"`
	SE92 *Reserved
	SE93 *N  `format:"" length:"4" validator:"N"`
	SE94 *N  `format:"LLVAR" length:"19" validator:"N"`
	SE95 *AN `format:"" length:"2" validator:"AN"`
	SE96 *N  `format:"LLVAR" length:"19" validator:"N"`
	SE97 *N  `format:"" length:"11" validator:"N"`
	SE98 *AN `format:"" length:"1" validator:"AN"`
	SE99 *AN `format:"LLVAR" length:"99" validator:"AN"`
}

func (m *SubMessage) Encode() ([]byte, error) {
	res := make([]byte, 0)

	// initialize bitmaps
	var bitmapPrimary uint64
	var bitmapSecondary uint64

	data := make([]byte, 0, 512)

	v := reflect.Indirect(reflect.ValueOf(m))
	t := v.Type()
	// iterate through iso fields, if field is not empty,
	// encode and append it to data, and set the proper bit in bitmap
	for i := 0; i < v.NumField(); i++ {
		if v.Field(i).Kind() != reflect.Ptr || v.Field(i).IsNil() {
			continue
		}
		sf := t.Field(i)
		// get field index, e.g. for SE2 index=2
		index, err := strconv.Atoi(strings.Trim(sf.Name, "SE"))
		if err != nil {
			return nil, err
		}
		// field length
		length, err := strconv.Atoi(sf.Tag.Get("length"))
		if err != nil {
			return nil, err
		}
		// iso field format, "" means that field has fixed length, but
		// e.g. LLVAR means that length indicator is encoded in the first 2 byte
		// in this case length means "maximum length
		format := sf.Tag.Get("format")
		validator := sf.Tag.Get("validator")

		f := v.Field(i).Interface().(field)

		if index <= 64 {
			bitmapPrimary = addField(bitmapPrimary, uint8(index))
		} else {
			// if we need secondary bitmap, set first bit in primary bitmap
			bitmapPrimary |= 1 << 63
			bitmapSecondary = addField(bitmapSecondary, uint8(index-64))
		}

		// encode field, append it to data
		d, err := f.Encode(m.encoder, length, format, validator)
		if err != nil {
			return nil, err
		}
		data = append(data, d...)
	}

	// append bitmaps to result
	m.bitmapPrimary = bitmapPrimary
	res = append(res, []byte(bitmapHex(bitmapPrimary))...)

	if bitmapSecondary != 0 {
		m.SE1 = bitmapSecondary
		res = append(res, []byte(bitmapHex(bitmapSecondary))...)
	}

	// append iso data elements to result
	res = append(res, data...)
	return res, nil
}

func (m *SubMessage) Decode(bytes []byte) error {
	var err error

	// decode MTI

	// it is an iterator, watching where we are currently in the iteration,
	// which byte will be the starting position of the next decode
	it := 0

	// decode bitmaps
	//decode primary bitmap
	m.bitmapPrimary, err = decodeHexString(string(bytes[it : it+16]))
	if err != nil {
		return err
	}
	it += 16

	// if first bit is 1, it means that we have secondary bitmap, decode secondary bitmap
	if isBitSet(m.bitmapPrimary, 1) {
		m.SE1, err = decodeHexString(string(bytes[it : it+16]))
		if err != nil {
			return err
		}
		it += 16
	}

	v := reflect.Indirect(reflect.ValueOf(m))
	t := v.Type()
	// iterate through iso fields, if bitmap is not empty at bit position i,
	// set field with index i with proper value
	for i := 0; i < v.NumField(); i++ {
		if v.Field(i).Kind() != reflect.Ptr {
			continue
		}
		sf := t.Field(i)
		// get field index, e.g. for DE2 index=2
		index, err := strconv.Atoi(strings.Trim(sf.Name, "SE"))
		if err != nil {
			return err
		}
		// if index < 64, search in primary bitmap if it is set
		if index <= 64 && !isBitSet(m.bitmapPrimary, uint8(index)) {
			continue
		}

		// if index > 64, search in secondary bitmap if it is set
		if index > 64 && !isBitSet(m.SE1, uint8(index-64)) {
			continue
		}

		// field (maximum) length
		length, err := strconv.Atoi(sf.Tag.Get("length"))
		if err != nil {
			return err
		}
		// iso field format, "" means that field has fixed length, but
		// e.g. LLVAR means that length indicator is encoded in the first 2 byte
		// in this case length means "maximum length
		format := sf.Tag.Get("format")
		validator := sf.Tag.Get("validator")

		// Decode field
		structField := v.Field(i)

		// skip if unexported
		if !structField.CanSet() {
			continue
		}
		// initialize field with empty struct
		fieldTyp := reflect.New(structField.Type().Elem())
		structField.Set(fieldTyp)

		f := v.Field(i).Interface().(field)
		nextFieldOffset, err := f.Decode(bytes[it:], m.encoder, length, format, validator)
		if err != nil {
			return err
		}
		it += nextFieldOffset
	}
	return nil
}
